### ISSUE TITLE
UnexpectedValueException in templates chain

### ISSUE THREAD

**Issue by sserbin** [vimeo/psalm#816]
related to (I guess): https://github.com/vimeo/psalm/issues/794

test case:
```php
<?php
/** @template T */
class Foo
{
    /**
     * @psalm-var class-string
     */
    private $type;

    /** @var array<T> */
    private $items;

    /**
     * @param class-string $type
     * @template-typeof T $type
     */
    public function __construct(string $type)
    {
        if (!in_array($type, [A::class, B::class], true)) {
            throw new \InvalidArgumentException;
        }
        $this->type = $type;
        $this->items = [];
    }

    /** @param T $item */
    public function add($item): void
    {
        $this->items[] = $item;
    }
}
// ^ same as in #794

class FooFacade
{
    /**
     * @template T
     * @param  T $item
     */
    public function add($item): void
    {
        $foo = $this->ensureFoo([$item]);
        $foo->add($item);
    }

    /**
     * @template T
     * @param  array<mixed,T> $items
     * @return Foo<T>
     */
    private function ensureFoo(array $items): EntitySeries
    {
        $type = $items[0] instanceof A ? A::class : B::class;
        return new Foo($type);
    }
}

class A {}
class B {}
```

expected: no errors
actual:
```
PHP Fatal error:  Uncaught UnexpectedValueException: There must be more than one unique type in /home/user/.config/composer/vendor/vimeo/psalm/src/Psalm/Checker/TypeChecker.php:1119
Stack trace:
#0 /home/user/.config/composer/vendor/vimeo/psalm/src/Psalm/Checker/Statements/Expression/CallChecker.php(1470): Psalm\Checker\TypeChecker::simplifyUnionType(Object(Psalm\Codebase), Object(Psalm\Type\Union))
#1 /home/user/.config/composer/vendor/vimeo/psalm/src/Psalm/Checker/Statements/Expression/CallChecker.php(908): Psalm\Checker\Statements\Expression\CallChecker::checkFunctionArgumentType(Object(Psalm\Checker\StatementsChecker), Object(Psalm\Type\Union), Object(Psalm\Type\Union), 'Foo::add', 0, Object(Psalm\CodeLocation), Object(PhpParser\Node\Expr\Variable), Object(Psalm\Context), false)
#2 /home/user/.config/composer/vendor/vimeo/psalm/src/Psalm/Checker/Statements/Expression/CallChecker.php(225): Psalm\Checker\Statements\Expression\CallChecker::checkFunctionLikeArgumentsMatch(Object(Psalm\Checker\StatementsChecker), Array, in /home/user/.config/composer/vendor/vimeo/psalm/src/Psalm/Checker/TypeChecker.php on line 1119

```

**Comment by muglug** [comment]
Fixed in dc02ca8

**Comment by muglug** [comment]
Thanks, that was an edge-case I hadn't thought about
